File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int top = 0;
        int bottom = matrix.length-1;
        int row = -1;
        
        while(top<=bottom){
            int mid = top + (bottom-top)/2;
            if(matrix[mid][0] > target){
                bottom = mid-1;
            }else{
                row = mid;
                top = mid+1;
            }
        }
        
        if(row==-1){
            return false;
        }
        
        int start = 0;
        int end = matrix[0].length -1;
        
        while(start<=end){
            int mid = start + (end-start)/2;
            
            if(matrix[row][mid]==target){
                return true;
            }
            
            if(matrix[row][mid] > target){
                end = mid-1;
            }else{
                start = mid+1;
            }
        }
        
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])
        i = 0
        j = m - 1
        
        while(i>=0 and i<n and j>=0 and j<m):
    
            if matrix[i][j] == target:
                return True
    
            if target < matrix[i][j]:
                j-=1
            else:
                i+=1
        
        return False